xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,201 @@
+func createClusterNode(clusterName string, xdsClient xdsClientInterface, topLevelHandler *clusterHandler) *clusterNode {
+	print("|createClusterNode() called on ", clusterName, "|")
+	c := &clusterNode{
+		clusterHandler: topLevelHandler,
+	}
+	print("About to communicate with the xds client for:", clusterName, "|")
+	// Communicate with the xds client here.
+	c.cancelFunc = xdsClient.WatchCluster(clusterName, c.handleResp)
+	return c
+}
+
+// This function cancels the cluster watch on the cluster and all of it's children.
+func (c *clusterNode) delete() {
+	c.cancelFunc()
+	for _, child := range c.children {
+		child.delete()
+	}
+}
+
+// Construct cluster update (potentially a list of ClusterUpdates) for a node.
+func (c *clusterNode) constructClusterUpdate() ([]xdsclient.ClusterUpdate, error) { // THIS LOGIC IS WRONG
+	print("Tries constructing a cluster update for Cluster ", c.clusterUpdate.ServiceName, ". ")
+	// If the cluster has not yet received an update, the cluster update is not yet ready.
+	if !c.receivedUpdate {
+		return nil, errors.New("Tried to construct a cluster update on a cluster that has not received an update.")
+	}
+
+	// Base case - LogicalDNS or EDS. Both of these cluster types will be tied to a single ClusterUpdate.
+	if c.clusterUpdate.ClusterType != xdsclient.ClusterTypeAggregate {
+		print("Returning the cluster update for service (base node):", c.clusterUpdate.ServiceName)
+		return []xdsclient.ClusterUpdate{c.clusterUpdate}, nil
+	}
+
+	// If an aggregate construct a list by recursively calling down to all of it's children.
+	var childrenUpdates []xdsclient.ClusterUpdate
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		childrenUpdates = append(childrenUpdates, childUpdateList...)
+	}
+	print("Returning the update list for an aggregate cluster")
+	return childrenUpdates, nil
+}
+
+// handleResp handles a xds response for a particular cluster. This function also
+// handles any logic with regards to any child state that may have changed.
+// Problem statement: perhaps this logic is wrong
+func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err error) {
+	c.clusterHandler.clusterMutex.Lock()
+	defer c.clusterHandler.clusterMutex.Unlock()
+	print("SERVICE NAME:", clusterUpdate.ServiceName)
+	if err != nil { // Write this error for run() to pick up in CDS LB policy.
+		c.clusterHandler.updateChannel<-clusterHandlerUpdate{chu: nil, err: err}
+		return
+	}
+	// This variable will determine whether there was a delta with regards to this clusterupdate. If there was, at the end
+	// of the response ping ClusterHandler to send CDS Policy a new list of ClusterUpdates.
+	var delta bool
+	// If this is the first update to the Cluster Node, ping ClusterHandler at the end of method to try and build a config.
```

Comment:
This is opposite to what I think. If there's a change in an aggregate node, there's no need to notify the top handler, because the one of the children will change, and this child will need to wait for an update.

I think we can get away without optimizing this. Even if we call the top handler for each update, the recursive function will return error when it reaches the new child.

If you want to optimize, I think the only case to notify the top handler is when a leaf node receives an update that's different from the previous.

Reply:
Hmmmmmm...not just receivedUpdate, but also just the delta logic in general. Seems illogical to have all this logic about delta when children will not have received update. Maybe it would be better to ping on any update to leaf node.

Reply:
I want to discuss this with you right now but problem is my shift started 7 hours before yours lol. TIME ZONES

Reply:
Ok, this took some thought. I settled on this logic: if an aggregate, ping on whether a child was deleted AND also no children were created. If a leaf, ping on whether delta in clusterUpdate. This logic is implemented now.